### ISSUE TITLE
Custom ASCII Art

### ISSUE THREAD

**Issue by NavTech270** [Gobidev/pfetch-rs#33]
How do you add custom ASCII art when you install `pfetch-rs` via `Homebrew`? I want to use another fedora logo from `ufetch`. I've checked in the cellar from brew to see if there are any files that contain ASCII art but there is none. There is not even a single file from this git repository in the Brew Cellar. I might not know how Homebrew works properly? @Gobidev

**Comment by Gobidev** [comment]
Currently, the only way to supply custom ASCII art is replacing the logo in `pfetch-extractor/logos.sh` and rebuilding the project. I don't really know how Homebrew works and how you would edit the logos there, however I am considering adding support for supplying custom logos at runtime, which would make this a lot easier.

**Comment by NavTech270** [comment]
Would this make customizing with a Homebrew installation possible? @Gobidev

**Comment by Gobidev** [comment]
Yes, it would allow for customization with any type of installation.

**Comment by NavTech270** [comment]
Will you be releasing this soon? @Gobidev

**Comment by NavTech270** [comment]
I see, I may resort to a manual installation rather than Homebrew for now I guess to use custom ASCII art @Gobidev

**Comment by jadelily18** [comment]
would it be worth moving everything configurable into something like  `~/.config/pfetch/config.toml`? i don't really see the benefit of having everything in environment variables

**Comment by OldWorldOrdr** [comment]
The reason is to maintain compatibility with old pfetch configs

**Comment by jadelily18** [comment]
> The reason is to maintain compatibility with old pfetch configs

ah i see. would it be worth moving everything in a config file but somehow keeping old configuration options as well? i feel like it would be easier to add custom ascii in a config file than in an environment variable

**Comment by RubixDev** [comment]
@jadelily18 you can always set `PF_SOURCE` to a file that acts as a config file, containing all the other environment variables. Despite the README saying that the specified file can be any valid shell script, it must only contain env var declarations, comments, and whitespace (such as [the code block in the README](https://github.com/Gobidev/pfetch-rs#configuration))

**Comment by jadelily18** [comment]
> @jadelily18 you can always set `PF_SOURCE` to a file that acts as a config file, containing all the other environment variables. Despite the README saying that the specified file can be any valid shell script, it must only contain env var declarations, comments, and whitespace (such as [the code block in the README](https://github.com/Gobidev/pfetch-rs#configuration))

oh fair that makes sense thank you!

**Comment by NavTech270** [comment]
Do you mean a config file like `~/.config/pfetch-rs/config.toml`? @RubixDev

**Comment by RubixDev** [comment]
Sort of, but it is not actually in TOML format and you must specify the file location yourself (using `PF_SOURCE`)

**Comment by OldWorldOrdr** [comment]
i have mine set to `~/.config/pfetch.conf`

**Comment by OldWorldOrdr** [comment]
```sh
# vim:filetype=sh
# pfetch configuration file
PF_INFO="ascii title os host kernel uptime pkgs"
PF_COL3="COL1"
```
this is all it is

**Comment by Gobidev** [comment]
Thanks to the help of @RubixDev, I refactored the logo-parsing code to enable runtime parsing of logos with the same codebase that is used for compile-time parsing, the progress can be found at #34. With that, it is not much more work to support custom logos at runtime.

**Comment by Gobidev** [comment]
Custom logos at runtime are now supported as of version 2.8.0. The logos need to be put into a file in the pfetch logo format (see `README.md` for details) and loaded with the `PF_CUSTOM_LOGOS` environment variable. You can add new logos and/or overwrite exisiting logos with your own.

**Comment by NavTech270** [comment]
@Gobidev @RubixDev I have re-installed pfetch-rs via Homebrew and I still can see no config files. Do I have to manually create these myself? None of the files from this repository appear in my `~/.config/`. I tested installing neofetch via Homebrew and it automatically made a `config.conf` file for me within `~/.config/neofetch`. I assumed I'd have access to folders like pfetch_extractor somewhere or pfetch-rs wouldn't work on my system? I've tried researching about Homebrew but didn't get any luck so was wondering when you setup the installation for it, do you know where the files from this repository are stored when installing via `brew`?

**Comment by RubixDev** [comment]
@NavTech270 pfetch does not create config files by itself, the new version doesn't change that. But, as I already explained, you can set a custom path to a custom config file using the `PF_SOURCE` variable. You don't need to do that for custom logos though. For that you instead use the new `PF_CUSTOM_LOGOS` variable which points to another file which contains your custom logos. An example for such a file is in the README.

If you just don't know how to set environment variables, there are multiple ways, but usually your shell has a file that is sourced when entering the shell (like `~/.bashrc` for bash or `~/.zshrc` for zsh) and in there you can set your environment variables using `export VAR=value` (so for example `export PF_CUSTOM_LOGOS=~/.config/pfetch/logos.sh`).

**Comment by NavTech270** [comment]
Got it! Just to clarify, what could an example config file look like for `PF_SOURCE`? I assume by the README, it also has to have an extension of `.sh` and not `.conf` like pfetch? @RubixDev

**Comment by Gobidev** [comment]
> Got it! Just to clarify, what could an example config file look like for PF_SOURCE? I assume by the README, it also has to have an extension of .sh and not .conf like pfetch? @RubixDev

The extension does not matter, it could look something like this:
```sh
PF_INFO="title os host kernel uptime pkgs memory cpu"
PF_ASCII="openbsd"
PF_COL3=COL1
```

**Comment by NavTech270** [comment]
It works! Thanks for adding custom ASCII art feature @RubixDev @Gobidev

**Comment by NavTech270** [comment]
Do you have any efficient ways of making the custom art itself if I wanted to design my won and put it in the `PF_CUSTOM_LOGOS`? @RubixDev @Gobidev

**Comment by Gobidev** [comment]
> Do you have any efficient ways of making the custom art itself if I wanted to design my won and put it in the PF_CUSTOM_LOGOS? @RubixDev @Gobidev

Not really, I normally paste the logo I want to create into Gimp and try to follow the shape with a monospace font, but it takes a lot of trial and error.

**Comment by NavTech270** [comment]
I've seen ways where you can convert a JPG/PNG into ASCII format but I assume that's not possible here? @Gobidev

**Comment by Gobidev** [comment]
> I've seen ways where you can convert a JPG/PNG into ASCII format but I assume that's not possible here?

You can certainly try, but the tools that exist probably don't create good looking output for such a small size (<10 rows).

**Comment by NavTech270** [comment]
Thanks anyway @Gobidev
